### PR TITLE
drivers: sensor: bosch: add support for bmp180

### PR DESCRIPTION
**Pull request by jovicjakov**
This PR aims to add driver support for the Bosch BMP180 digital pressure and temperature sensor.

### CODE REVIEW

File: samples/sensor/bmp180/src/main.c
Comment (kartben) [inline samples/sensor/bmp180/src/main.c:1]: please just drop this sample altogether as it's not adding anything useful compared to e.g. sensor_shell sample which folks can use to quickly test their sensor.

**Review by kartben — DISMISSED** [review]
some early comments. thanks for this new driver!

File: drivers/sensor/bosch/bmp180/bmp180.c [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+}
+
+static int bmp180_read_uncomp_temp(const struct device *dev)
+{
+	const struct bmp180_config *config = dev->config;
+	struct bmp180_data *data = dev->data;
+	uint8_t buf[2];
+
+	if (i2c_reg_write_byte_dt(&config->i2c, BMP180_REG_CTRL_MEAS, BMP180_CMD_READ_TEMP) < 0) {
+		LOG_ERR("Failed to start the temperature measurement.");
```
[the remark this answers is not in the document]
Comment (jovicjakov) [reply drivers/sensor/bosch/bmp180/bmp180.c:60]: will do

File: drivers/sensor/bosch/bmp180/bmp180.c [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+		val->val2 = data->temp % 10 * 100000;
+		break;
+	case SENSOR_CHAN_PRESS:
+		/* Pressure is calculated in steps of 1 Pa */
+		val->val1 = data->press;
+		val->val2 = 0;
+		break;
+	case SENSOR_CHAN_ALTITUDE:
+		val->val1 = data->alt;
+		val->val2 = 0;
```
[the remark this answers is not in the document]
Comment (jovicjakov) [reply drivers/sensor/bosch/bmp180/bmp180.c:244]: good catch, thanks

File: samples/sensor/bmp180/src/main.c
Comment (jovicjakov) [reply samples/sensor/bmp180/src/main.c:1]: I agree, I'll remove the commit

**Review by decsny — DISMISSED** [review]

**Review by ubieda — COMMENTED** [review]
Thanks for your contribution! Small suggestion, otherwise looks good!

File: drivers/sensor/bosch/bmp180/bmp180.c [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+		return -EINVAL;
+	}
+
+	/* Read the MSB, LSB, and XLSB from the sensor */
+	if (i2c_burst_read_dt(&config->i2c, BMP180_REG_OUT_MSB, buf, sizeof(buf)) < 0) {
+		LOG_ERR(ERROR_MSG_FMT, "read", "pressure measurement");
+		return -EIO;
+	}
+
+	data->uncomp_press = ((int32_t)buf[0] << 16) | ((int32_t)buf[1] << 8) | buf[2];
```
[the remark this answers is not in the document]
Comment (jovicjakov) [reply drivers/sensor/bosch/bmp180/bmp180.c:115]: ofc, I'll refactor

**Review by ubieda — DISMISSED** [review]

**Comment by jovicjakov** [comment]
@decsny, when you have time, could you take another look at this? Thanks!

**Review by decsny — DISMISSED** [review]

File: drivers/sensor/bosch/bmp180/bmp180.c [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+	}
+	return 0;
+}
+
+static const struct sensor_driver_api bmp180_driver_api = {
+	.sample_fetch = bmp180_sample_fetch,
+	.channel_get = bmp180_channel_get,
+};
+
+int bmp180_init(const struct device *dev)
```
Comment (MaureenHelm) [inline drivers/sensor/bosch/bmp180/bmp180.c:253]:
```suggestion
static int bmp180_init(const struct device *dev)
```

File: drivers/sensor/bosch/bmp180/bmp180.c [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+{
+	const struct bmp180_config *config = dev->config;
+	uint8_t id = 0U;
+
+	if (!device_is_ready(config->i2c.bus)) {
```
Comment (MaureenHelm) [inline drivers/sensor/bosch/bmp180/bmp180.c:258]:
```suggestion
	if (!i2c_is_ready_dt(&config->i2c)) {
```

File: drivers/sensor/bosch/bmp180/Kconfig [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+# Copyright (C) 2024 Jakov Jovic <jovic.jakov@outlook.com>
+# SPDX-License-Identifier: Apache-2.0
+
+config BMP180
+	bool "Bosch BMP180 digital pressure and temperature sensor"
+	depends on DT_HAS_BOSCH_BMP180_ENABLED
+	select I2C
+	default y
+	help
+		Enable the driver for the Bosch BMP180 pressure sensor.
```
Comment (MaureenHelm) [inline drivers/sensor/bosch/bmp180/Kconfig:12]:
```suggestion
	  Enable the driver for the Bosch BMP180 pressure sensor.
```

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Some nits, but otherwise looks good. Thanks for your contribution!

File: drivers/sensor/bosch/bmp180/bmp180.c (hunk 4 above)
Comment (jovicjakov) [reply drivers/sensor/bosch/bmp180/bmp180.c]: Thanks, resolved

File: drivers/sensor/bosch/bmp180/bmp180.c (hunk 5 above)
Comment (jovicjakov) [reply drivers/sensor/bosch/bmp180/bmp180.c]: Thanks, resolved

File: drivers/sensor/bosch/bmp180/Kconfig (hunk 6 above)
Comment (jovicjakov) [reply drivers/sensor/bosch/bmp180/Kconfig]: Thanks, resolved

**Comment by jovicjakov** [comment]
@MaureenHelm Thank you for the review! I've fixed the issues you pointed out. Please take another look when you have a chance.  :)

**Comment by ubieda** [comment]
@MaureenHelm it does appear everything's addressed. There's a conflict resolution required but seems trivial to fix.

**Review by ubieda — DISMISSED** [review]

**Comment by kartben** [comment]
> @MaureenHelm it does appear everything's addressed. There's a conflict resolution required but seems trivial to fix.

@ubieda  feel free to directly push to the branch to save some time :)

**Review by ubieda — DISMISSED** [review]
Re-approving after rebasing

**Review by teburd — DISMISSED** [review]

**Comment by aescolar** [comment]
Bulk changing the milestone from everything which did not make it to the v3.7.0 freeze deadline
and which is not labeled as a bug or documentation (or does not appear clearly as only containing
a bug fix or a documentation change).
If you disagree about this change please add the milestone again.

Please note that until rc2 only PRs containing bug fixes, docs and tests can be merged.
After that and until release only bug fixes and docs changes.

Exceptions require approval by the release team and a vote by the TSC.
https://docs.zephyrproject.org/latest/project/release_process.html#stabilization-phase

**Comment by aescolar** [comment]
@jovicjakov rebase needed

**Comment by kartben** [comment]
rebased to resolve conflict

**Review by ubieda — APPROVED** [review]

**Review by kartben — APPROVED** [review]

**Comment by aescolar** [comment]
@MaureenHelm please re-review

**Comment by mariopaja** [comment]
This driver can also be used on BMP085. Maybe it can be documented somewhere since https://github.com/zephyrproject-rtos/zephyr/pull/67612 was blocked

File: drivers/sensor/bosch/bmp180/bmp180.c [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+	struct bmp180_data *data = dev->data;
+
+	switch (chan) {
+	case SENSOR_CHAN_AMBIENT_TEMP:
+		/* Temperature is calculated in steps of 0.1 °C*/
+		val->val1 = data->temp / 10;
+		val->val2 = data->temp % 10 * 100000;
+		break;
+	case SENSOR_CHAN_PRESS:
+		/* Pressure is calculated in steps of 1 Pa */
```
Comment (ruehlchris) [inline drivers/sensor/bosch/bmp180/bmp180.c:234]: CHAN PRESSURE for the Senor API should be in kPa not Pa.

File: drivers/sensor/bosch/bmp180/bmp180.c [hunk 8]
```diff
@@ -0,0 +1,294 @@
...
+
+	/* See datasheet, page 15 for more details */
+	int32_t x1 = (data->uncomp_temp - (int32_t)data->ac6) * ((int32_t)data->ac5) >> 15;
+	int32_t x2 = ((int32_t)data->mc << 11) / (x1 + (int32_t)data->md);
+	int32_t b5 = x1 + x2;
+
+	int32_t b6 = b5 - 4000;
+
+	x1 = (data->b2 * ((b6 * b6) >> 12)) >> 11;
+	x2 = (data->ac2 * b6) >> 11;
```
Comment (ruehlchris) [inline drivers/sensor/bosch/bmp180/bmp180.c:159]: you might want to simplify from the original data-sheet example to 
x1 = ((b2 * (b6 * b6)>>12))>>11;
x1 = ((b2 * b6 * b6) / pow(2,12) )/ pow(2,11)
x1 = (b2 * b6 *b6) (pow(2,12) * pow(2,11))
x1 = (b2 * b6 * b6) / 0x800000;  << much

File: drivers/sensor/bosch/bmp180/bmp180.c [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+
+	uint32_t b7 = ((uint32_t)data->uncomp_press - b3) * (50000 >> config->oversampling);
+	int32_t p;
+
+	if (b7 < 0x80000000) {
+		p = (b7 * 2) / b4;
+	} else {
+		p = (b7 / b4) * 2;
+	}
+
```
Comment (ruehlchris) [inline drivers/sensor/bosch/bmp180/bmp180.c:176]: I wondered the if (b7< xx) statement here from the datasheet example
math wise it is exactly same  only a data overflow can happen if b7 is too big
so b7/b4 *2 would always be safe and no need for a if()

**Comment by ruehlchris** [comment]
Hi @jovicjakov ,  it is a pity that I missed you PR request and did my own implementation. 
Not a problem but wast a bit of time.  I saw in your implementation a lot of the original code from the datasheet which are not optimal (IMHO). I add some comments to your code lately. 
As well, you exit with an error if the command ready bit(5) not 0 rather wait for the convention is ready.
The Sensor API wants the Pressure return as kPa your implementation return Pa.
Please have a look into my PR and use it to your benefit.  #77182

**Comment by jovicjakov** [comment]
Hi everyone,

I just wanted to inform you that I have started an internship recently, and due to the terms of my contract, I am not allowed to contribute to open-source projects, not at this moment at least. I won't be able to continue working on this PR or addressing the remaining comments.

If any further changes are required, I may be unable to make them myself. I would be grateful if someone else could take over from here. Thanks!

**Pull request closed without merging**